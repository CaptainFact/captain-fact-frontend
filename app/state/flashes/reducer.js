import { List, Record } from 'immutable'
import { handleActions, createAction, combineActions } from 'redux-actions'
import { buildFlash } from './record'
import { getErrorInfo } from '../../lib/errors'
import { NO_INTERNET_ERROR } from '../../constants'


// Actions
export const addFlash = createAction('FLASHES/ADD', options => buildFlash(options))
export const removeFlash = createAction('FLASHES/REMOVE')
export const pause = createAction('FLASHES/PAUSE', () => true)
export const unPause = createAction('FLASHES/UNPAUSE', () => false)
export const update = createAction('FLASHES/UPDATE')

// Actions helpers (use them like regular actions)
export const flashError = options => addFlash({
  flashType: 'danger',
  iconName: 'exclamation-circle',
  ...options
})
export const flashErrorMsg = message => flashError({message})
export const flashErrorUnauthenticated = () => flashError({
  message: 'errors:server.unauthenticated',
  infoText: 'main:menu.loginSignup',
  infoUrl: '/login'
})

export const flashSuccessMsg = (message, params = {}) => addFlash({
  flashType: 'success',
  iconName: 'check-circle',
  message,
  ...params
})
export function errorToFlash(msg) {
  const errorInfo = getErrorInfo(msg)
  let action = null
  if (!errorInfo)
    action = flashError({message: msg, isError: true})
  else
    action = flashError({
      message: msg,
      infoUrl: errorInfo.url,
      infoText: errorInfo.i18nKey,
      isError: true
    })
  action.error = true
  return action
}
// Same as errorToFlash but doesn't show anything if payload is
// not a string (useful for forms)
export function errorMsgToFlash(msg) {
  if (typeof (msg) === 'string')
    return errorToFlash(msg)
  return () => () => null
}

// Reducer
const INITIAL_STATE = Record({
  flashes: new List(),
  isPaused: false
})

const FlashesReducer = handleActions({
  [addFlash]: (state, {payload}) => {
    // Only display one error for connections problems (instead of one per request)
    if (payload.message === NO_INTERNET_ERROR
      && state.flashes.find(f => f.message === NO_INTERNET_ERROR))
      return state
    return state.update('flashes', l => l.push(payload))
  },
  [removeFlash]: (state, {payload: {id}}) => {
    const flashIdx = state.flashes.findIndex(msg => msg.id === id)
    if (flashIdx !== -1)
      return state.update('flashes', l => l.delete(flashIdx))
    return state
  },
  [combineActions(pause, unPause)]: (state, {payload}) => state.set('isPaused', payload),
  [update]: (state, {payload}) => {
    if (!state.isPaused)
      return state.update('flashes', flashes => flashes
        .map(f => f.set('timeLeft', f.timeLeft - payload))
        .filter(msg => msg.timeLeft > 0)
      )
    return state
  }
}, INITIAL_STATE())

export default FlashesReducer
