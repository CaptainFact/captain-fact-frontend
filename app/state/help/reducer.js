import { Record } from "immutable"
import { createAction, handleActions } from 'redux-actions'


export const setLoading = createAction('HELP/SET_LOADING')
export const reset = createAction('HELP/RESET')
export const setContent = createAction('HELP/SET_HELP_PAGE')

const INITIAL_STATE = new Record({
  markdownContent: "",
  isLoading: false,
  error: null
})

const HelpReducer = handleActions({
  [setContent]: {
    next: (state, {payload}) => state.merge({
      markdownContent: payload,
      isLoading: false
    }),
    throw: (state, {payload}) => state.merge({error: payload, isLoading: false})
  },
  [setLoading]: (state, {payload}) => state.set('isLoading', payload),
  [reset]: () => INITIAL_STATE()
}, INITIAL_STATE())

export default HelpReducer
