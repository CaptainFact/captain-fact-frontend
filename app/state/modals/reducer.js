import { Stack } from 'immutable'
import { createAction, handleActions } from 'redux-actions'


export const addModal = createAction('ADD_MODAL')
export const popModal = createAction('POP_MODAL')

const ModalsReducer = handleActions({
  [addModal]: (state, {payload: {Modal, props}}) => state.push({Modal, props}),
  [popModal]: state => state.pop()
}, new Stack())

export default ModalsReducer
