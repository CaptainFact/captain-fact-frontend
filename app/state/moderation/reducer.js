import { Record, List, Map } from "immutable"
import { createAction, handleActions } from 'redux-actions'
import parseDateTime from '../../lib/parse_datetime'

import UserAction from '../user_actions/record'


export const setLoading = createAction('MODERATION/SET_LOADING')
export const setItems = createAction('MODERATION/SET_ITEMS')
export const removeItem = createAction('MODERATION/REMOVE_ITEM')

const INITIAL_STATE = new Record({
  isLoading: false,
  error: null,
  items: new List()
})

const ModerationReducer = handleActions({
  [setItems]: {
    next: (state, {payload}) => state.merge({
      items: new List(payload.map(prepareAction)),
      isLoading: false
    }),
    throw: (state, action) => state.merge({
      isLoading: false,
      error: action.payload
    })
  },
  [setLoading]: (state, {payload}) => state.set('isLoading', payload),
  [removeItem]: (state, {payload}) => {
    const moderationIdx =
      state.items.findIndex(moderationEntry => moderationEntry.id === payload)
    if (moderationIdx !== -1)
      return state.update('items', i => i.delete(moderationIdx))
  }
}, INITIAL_STATE())

const prepareAction = (action) => {
  action.time = parseDateTime(action.time)
  action.changes = new Map(action.changes)
  return UserAction(action)
}

export default ModerationReducer
