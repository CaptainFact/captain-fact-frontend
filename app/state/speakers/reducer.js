import { Record, List } from 'immutable'
import { createAction, handleActions } from 'redux-actions'
import Speaker from './record'


export const setLoading = createAction('SPEAKERS/SET_LOADING')
export const setLoadingWiki = createAction('SPEAKERS/SET_LOADING_WIKI')
export const reset = createAction('SPEAKERS/RESET')
export const actionFetchSpeaker = createAction('SPEAKERS/FETCH_ONE')
export const actionFetchSpeakerWikiLinks = createAction('SPEAKERS/FETCH_WIKI_DATA')

const supportedSites = ['wikimedia', 'wikipedia', 'wikiquote', 'wikinews']


const INITIAL_STATE = new Record({
  currentSpeaker: new Speaker(),
  currentSpeakerSummary: "",
  currentSpeakerLinks: new Record({
    wikimedia: null,
    wikipedia: null,
    wikiquote: null,
    wikinews: null
  })(),
  isLoading: false,
  isLoadingWiki: false,
  error: null
})

const SpeakersReducer = handleActions({
  [actionFetchSpeaker]: {
    next: (state, {payload}) => state.mergeDeep({isLoading: false, currentSpeaker: payload}),
    throw: (state, {payload}) => state.merge({isLoading: false, error: payload})
  },
  [actionFetchSpeakerWikiLinks]: {
    next: (state, {payload}) => {
      const allLinks = {}
      for (let link of payload) {
        const siteName = checkLink(link)
        if (siteName)
          allLinks[siteName] = link
      }
      return state.mergeDeep({isLoadingWiki: false, currentSpeakerLinks: allLinks})
    },
    throw: (state, {payload}) => state.merge({isLoadingWiki: false})
  },
  [setLoading]: (state, {payload}) => state.set('isLoading', payload),
  [setLoadingWiki]: (state, {payload}) => state.set('isLoadingWiki', payload),
  [reset]: () => INITIAL_STATE()
}, INITIAL_STATE())

function checkLink(url) {
  for (let siteName of supportedSites) {
    if (url.includes(siteName))
      return siteName
  }
  return null
}

export default SpeakersReducer