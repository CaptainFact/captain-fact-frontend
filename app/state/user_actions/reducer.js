import { Record, List, Map } from 'immutable'
import { createAction, handleActions, combineActions } from 'redux-actions'
import { diffWordsWithSpace } from 'diff'

import parseDateTime from '../../lib/parseDateTime'
import formatSeconds from "../../lib/seconds_formatter"
import UserAction from "./record"
import Statement from '../video_debate/statements/record'
import Speaker from '../speakers/record'
import {ENTITY_SPEAKER, ENTITY_STATEMENT} from '../../constants'
import { resetVideoDebate } from '../video_debate/actions'

export const setLoading = createAction('VIDEO_DEBATE_HISTORY/SET_LOADING')
export const reset = createAction('VIDEO_DEBATE_HISTORY/RESET')
export const fetchAll = createAction('VIDEO_DEBATE_HISTORY/FETCH')
export const addAction = createAction('VIDEO_DEBATE_HISTORY/ADD_ACTION')
export const generateDiff = createAction('VIDEO_DEBATE_HISTORY/GENERATE_DIFF')

const INITIAL_STATE = new Record({
  entitiesActions: new Map(),
  referenceEntities: new Map(),
  diffs: new Map(),
  isLoading: false,
  errors: null
})

const UsersActionsReducer = handleActions({
  [setLoading]: (state, {payload}) => state.set('isLoading', payload),
  [fetchAll]: {
    next: (state, {payload: {actions}}) => {
      const preparedActions = new List(actions.map(a => prepareAction(a)))
      const entitiesActions = preparedActions
        .sortBy(a => -a.time)
        .groupBy(a => entityKeyFromAction(a))
        .sortBy(actions => -actions.first().time)

      return state.merge({
        entitiesActions: entitiesActions,
        referenceEntities: buildReferenceEntities(entitiesActions),
        isLoading: false,
        errors: null
      })
    },
    throw: (state, {payload}) => state.merge({isLoading: false, errors: payload})
  },
  [addAction]: (state, {payload}) => {
    // /!\ Assumes that action is more recent than any other action previously stored
    const action = prepareAction(payload)
    const entityKey = entityKeyFromAction(action)
    return state.withMutations(state =>
      state
        .update('entitiesActions', e => e.withMutations(entitiesActions =>
          entitiesActions
            .update(entityKey, actions => actions.insert(0, action))
            .sortBy(actions => -actions.first().time)
        ))
        .updateIn(['referenceEntities', entityKey], reference =>
          buildReferenceEntity(new List([action]), reference)
        )
    )
  },
  [generateDiff]: (state, {payload}) => {
    // If diff already exists, do nothing
    if (state.diffs.has(payload.id))
      return state

    const entityActions = state.entitiesActions.get(`${payload.entity}:${payload.entity_id}`)
    const actionIdx = entityActions.findIndex(a => a.id === payload.id)
    // Get previous state
    let prevState = new Map()
    if (actionIdx + 1 < entityActions.size) {
      prevState = buildReferenceEntity(entityActions.slice(actionIdx + 1))
    }
    // Build changes object like key: [diffs]
    const diff = new Map().withMutations(diff => {
      for (let [key, newValue] of payload.changes.entrySeq()) {
        const valueDiff = diffEntry(key, prevState.get(key), newValue)
        diff.set(key, new List(valueDiff))
      }
    })
    return state.setIn(['diffs', payload.id], diff)
  },
  [combineActions(reset, resetVideoDebate)]: () => INITIAL_STATE()
}, INITIAL_STATE())
export default UsersActionsReducer


function completeReference(reference, actions, keysToStore) {
  return reference.withMutations(reference => {
    // Let's look for the most recent entries
    for (let action of actions) {
      if (action.changes.size === 0)
        continue
      for (let [idx, key] of keysToStore.entries()) {
        if (!(action.changes.has(key)))
          continue
        // Yihaa ! Changes contains a value for key
        reference.set(key, action.changes.get(key))
        delete(keysToStore[idx])
        if (keysToStore.length === 0)
          return reference
      }
    }
    return reference
  })
}

function buildReferenceStatement(actions, base=null) {
  if (!base)
    base = new Statement({ id: actions.last().entity_id, })
  return completeReference(base, actions, ['text', 'time', 'speaker_id'])
}

function buildReferenceSpeaker(actions, base=null) {
  if (!base)
    base = new Speaker({ id: actions.first().entity_id })
  return completeReference(base, actions, ['full_name', 'title'])
}

function buildReferenceEntity(actions, base=null) {
  const entity = actions.first().entity
  if (entity === ENTITY_STATEMENT)
    return buildReferenceStatement(actions, base)
  else if (entity === ENTITY_SPEAKER)
    return buildReferenceSpeaker(actions, base)
}

function buildReferenceEntities(entitiesActions) {
  return new Map().withMutations(references => {
    for (let [entity_key, actions] of entitiesActions.entrySeq()) {
      references.set(entity_key, buildReferenceEntity(actions))
    }
    return references
  })
}

function prepareAction(action) {
  action.time = parseDateTime(action.time)
  action.changes = new Map(action.changes)
  return UserAction(action)
}

function diffEntry(key, prevValue, newValue) {
  // Format numbers like prevNumber -> newNumber
  if (typeof(newValue) === "number") {
    // Format time like 0:42 -> 1:35
    if (key === "time") {
      prevValue = prevValue ? formatSeconds(prevValue) : ""
      newValue = formatSeconds(newValue)
    }
    // Generate diff
    if (prevValue)
      return [
        {removed: true, value: prevValue},
        {added: true, value: newValue}
      ]
    else
      return [{added: true, value: newValue}]
  }
  // Do a string diff
  return diffWordsWithSpace((prevValue || "").toString(), newValue ? newValue.toString() : "")
}

export function entityKeyFromAction(action) {
  return `${action.entity}:${action.entity_id}`
}


