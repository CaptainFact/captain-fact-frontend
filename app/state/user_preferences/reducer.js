import { Record } from 'immutable'
import { createAction, handleActions } from 'redux-actions'
import {
  ALL_VIDEOS,
  MOBILE_WIDTH_THRESHOLD,
  SUPPORTED_LOCALES
} from '../../constants'
import browserLocale from '../../i18n/browser_locale'

export const toggleSidebar = createAction('USER_PREFERENCES/TOGGLE_SIDEBAR')
export const closeSidebar = createAction('USER_PREFERENCES/CLOSE_SIDEBAR')
export const changeLocale = createAction('USER_PREFERENCES/CHANGE_LOCALE')
export const changeVideosLanguageFilter = createAction(
  'USER_PREFERENCES/CHANGE_VIDEOS_LANGUAGE_FILTER'
)
export const setVideosOnlyFromPartners = createAction(
  'USER_PREFERENCES/VIDEOS_ONLY_FROM_PATNERS'
)
export const toggleAutoscroll = createAction('STATEMENTS/TOGGLE_AUTOSCROLL')
export const toggleBackgroundSound = createAction(
  'STATEMENTS/TOGGLE_BACKGROUND_SOUND'
)

const isMobile = window.innerWidth <= MOBILE_WIDTH_THRESHOLD

const Preferences = new Record({
  // Disable autoscroll and sidebar expended by default on mobile
  sidebarExpended: !isMobile,
  locale: 'en',
  enableAutoscroll: !isMobile,
  enableSoundOnBackgroundFocus: true,
  videosLanguageFilter: null,
  videosOnlyFromPartners: ALL_VIDEOS
})

const loadState = () => {
  let localStoragePrefs = {}
  try {
    // Load preferences from localStorage
    localStoragePrefs = JSON.parse(localStorage.preferences)
  } catch (e) {
    // Or from default if it fails
    localStorage.preferences = JSON.stringify(Preferences())
    localStoragePrefs = { locale: browserLocale() }
  }
  // Ensure the locale is valid, use the default otherwise
  const preferences = Preferences().merge(localStoragePrefs)
  if (!SUPPORTED_LOCALES.includes(preferences.locale)) {
    return preferences.set('locale', browserLocale())
  }
  return preferences
}

const updateState = (state, key, value) => {
  state = state.set(key, value)
  localStorage.preferences = JSON.stringify(state.toJSON())
  return state
}

const UserPreferencesReducer = handleActions(
  {
    [toggleSidebar]: state => updateState(state, 'sidebarExpended', !state.sidebarExpended),
    [closeSidebar]: state => updateState(state, 'sidebarExpended', false),
    [changeLocale]: (state, { payload }) => updateState(state, 'locale', payload),
    [changeVideosLanguageFilter]: (state, { payload }) => updateState(state, 'videosLanguageFilter', payload),
    [setVideosOnlyFromPartners]: (state, { payload }) => updateState(state, 'videosOnlyFromPartners', payload),
    [toggleAutoscroll]: state => updateState(state, 'enableAutoscroll', !state.enableAutoscroll),
    [toggleBackgroundSound]: state => {
      return updateState(
        state,
        'enableSoundOnBackgroundFocus',
        !state.enableSoundOnBackgroundFocus
      )
    }
  },
  loadState()
)

export default UserPreferencesReducer
