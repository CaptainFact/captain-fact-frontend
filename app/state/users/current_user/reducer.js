import { Record } from "immutable"
import { createAction, handleActions } from 'redux-actions'

import User from "../record"

// Actions
export const set = createAction('CURRENT_USER/SET')
export const userLogin = createAction('CURRENT_USER/LOGIN')
export const setLoading = createAction('CURRENT_USER/SET_LOADING')
export const setPosting = createAction('CURRENT_USER/SET_POSTING')
export const reset = createAction('CURRENT_USER/RESET')

// Reducer

const INITIAL_STATE = new Record({
  error: null,
  isLoading: false,
  isPosting: false,
  data: new User()
})

const CurrentUserReducer = handleActions({
  [set]: {
    next: (state, {payload}) => {
      return state.merge({data: new User(payload) || {}, error: null, isLoading: false})
    },
    throw: (state, {payload}) => {
      return state.merge({error: payload, isLoading: false})
    }
  },
  [userLogin]: {
    next: (state, {payload}) => {
      return state.mergeDeep({data: payload || {}, error: null, isPosting: false})
    },
    throw: (state, {payload}) => {
      return state.merge({error: payload, isPosting: false})
    }
  },
  [setLoading]: (state, {payload}) => state.set('isLoading', payload),
  [setPosting]: (state, {payload}) => state.set('isPosting', payload),
  [reset]: () => INITIAL_STATE()
}, INITIAL_STATE())

export default CurrentUserReducer