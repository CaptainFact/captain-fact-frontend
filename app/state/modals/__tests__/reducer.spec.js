import Modal from '../../../components/Modal/Modal'
import reducer, { addModal, popModal } from '../reducer'

const INITIAL_STATE = reducer(undefined, {})

const createTestModal = (id) => ({
  Modal,
  props: {
    id,
    hello: 'world',
  },
})

it('should return the initial state', () => {
  snapshot(INITIAL_STATE)
})

it('should be able to add and remove modals', () => {
  snapshotReducer(
    reducer,
    INITIAL_STATE,
    addModal(createTestModal(1)),
    addModal(createTestModal(2)),
    popModal(),
    popModal(),
  )
})
