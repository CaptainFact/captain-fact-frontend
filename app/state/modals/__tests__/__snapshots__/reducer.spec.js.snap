// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be able to add and remove modals 1`] = `Immutable.Stack []`;

exports[`should be able to add and remove modals 2`] = `
Immutable.Stack [
  Object {
    "Modal": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "type": [Function],
    },
    "props": Object {
      "hello": "world",
      "id": 1,
    },
  },
]
`;

exports[`should be able to add and remove modals 3`] = `
Immutable.Stack [
  Object {
    "Modal": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "type": [Function],
    },
    "props": Object {
      "hello": "world",
      "id": 2,
    },
  },
  Object {
    "Modal": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "type": [Function],
    },
    "props": Object {
      "hello": "world",
      "id": 1,
    },
  },
]
`;

exports[`should be able to add and remove modals 4`] = `
Immutable.Stack [
  Object {
    "Modal": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "type": [Function],
    },
    "props": Object {
      "hello": "world",
      "id": 1,
    },
  },
]
`;

exports[`should be able to add and remove modals 5`] = `Immutable.Stack []`;

exports[`should return the initial state 1`] = `Immutable.Stack []`;
